[view-transitions] Support parsing and checking :active-view-transition-type()

tuankiet65 opened the pull request:
#### 44cfc58a80e6797fd128ce8bd4070a779c56980f
<pre>
[view-transitions] Support parsing and checking :active-view-transition-type()
<a href="https://bugs.webkit.org/show_bug.cgi?id=277779">https://bugs.webkit.org/show_bug.cgi?id=277779</a>
<a href="https://rdar.apple.com/133610087">rdar://133610087</a>

Reviewed by Tim Nguyen.

This patch adds support for parsing and checking the pseudo class
:active-view-transition-type(). This patch also modifies ViewTransitionTypeSet
to invalidate the stylesheet if a new type is added/removed. With this, support
for view transition types is completed, and is now enabled by default.

The WPT test changes are already upstreamed here:
<a href="https://github.com/web-platform-tests/wpt/commit/85b16fcfb299c9d6ff85dfe16beacda7fcb609d2">https://github.com/web-platform-tests/wpt/commit/85b16fcfb299c9d6ff85dfe16beacda7fcb609d2</a>

* LayoutTests/TestExpectations: Don&apos;t skip view transition type tests, now that they&apos;re passing.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-matches-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-matches-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-matches.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved-mutation-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved-mutation.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Enable ViewTransitionTypesEnabled by default
* Source/WebCore/css/CSSPseudoSelectors.json: added pseudo class :active-view-transition-type.
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::appendPossiblyQuotedIdentifier): renamed from appendLangArgument.
(WebCore::appendCommaSeparatedPossiblyQuotedIdentifierList): renamed from appendLangArgumentList.
(WebCore::CSSSelector::selectorText const): added code to serialize :active-view-transition-type.
(WebCore::appendLangArgument): Deleted.
(WebCore::appendLangArgumentList): Deleted.
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const): added code to check :active-view-transition-type.
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesActiveViewTransitionTypePseudoClass): added function to check :active-view-transition-type.
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h: added flag for parsing :active-view-transition-type.
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::consumeCommaSeparatedCustomIdentList): added function to parse a list of comma-separated custom idents (used by :active-view-transition-type)
(WebCore::CSSSelectorParser::consumePseudo): added code to parse :active-view-transition-type.
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h: added flag for parsing :active-view-transition-type.
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType): mark :active-view-transition-type as not JIT compilable.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setActiveViewTransition): when invalidating style, also invalidate :active-view-transition-type.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.h:
(WebCore::ViewTransition::setTypes): use WTFMove.
* Source/WebCore/dom/ViewTransitionTypeSet.cpp:
(WebCore::ViewTransitionTypeSet::clearFromSetLike): invalidate style on clearing.
(WebCore::ViewTransitionTypeSet::addToSetLike): invalidate style on adding new types.
(WebCore::ViewTransitionTypeSet::removeFromSetLike): invalidate style on deleting types.
(WebCore::ViewTransitionTypeSet::hasType const): renamed from has()
(WebCore::ViewTransitionTypeSet::has const): Deleted.
* Source/WebCore/dom/ViewTransitionTypeSet.h:

Canonical link: <a href="https://commits.webkit.org/282344@main">https://commits.webkit.org/282344@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63d41cf919edfe67118eb6d4eba664643eddc523

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15488 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66913 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13496 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65012 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13780 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50711 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9317 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65961 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39267 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54476 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31393 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35956 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11809 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12372 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/56002 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57504 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12139 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68608 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/62135 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6838 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11768 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58025 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6870 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54536 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58224 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13945 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5710 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83898 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38068 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14767 "Found 1 new JSC stress test failure: wasm.yaml/wasm/fuzz/memory.js.wasm-eager (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39148 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40259 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38890 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -374,17 +374,17 @@ static void appendPseudoClassFunctionTail(StringBuilder& builder, const CSSSelec
     if (!langArgument.wasQuoted)
         serializeIdentifier(langArgument.identifier, builder);
     else
         serializeString(langArgument.identifier, builder);
```
nit: rename `langArgument` as well.

nt1m commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -586,4 +587,24 @@ ALWAYS_INLINE bool matchesActiveViewTransitionPseudoClass(const Element& element
 
+ALWAYS_INLINE bool matchesActiveViewTransitionTypePseudoClass(const Element& element, const FixedVector<PossiblyQuotedIdentifier>& typesInSelector)
+{
+    // check if element is the root element
```
This comment is not necessary.

tuankiet65 replied:
I think it's obvious if you already know it, but I came in knowing nothing and it took me a while to realize it's a root check, so I want to leave a comment there for others.

nt1m replied:
`// This pseudo class only matches on the root element.` might be a better comment

tuankiet65 replied:
Sounds good!

nt1m commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -586,4 +587,24 @@ ALWAYS_INLINE bool matchesActiveViewTransitionPseudoClass(const Element& element
+    if (&element != element.document().documentElement())
+        return false;
+
+    if (const auto* viewTransition = element.document().activeViewTransition()) {
+        const auto& activeTypes = viewTransition->types();
+
+        for (const auto& type : typesInSelector) {
```
we don't typically use const auto, just auto

tuankiet65 replied:
It's to make sure we don't inadvertently modify them

nt1m commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -277,6 +277,46 @@ static FixedVector<PossiblyQuotedIdentifier> consumeLangArgumentList(CSSParserTo
+        return std::nullopt;
+
+    return token.value().toAtomString();
+}
```
I was going to suggest using `consumeCustomIdentRaw`, but I realize this is the selector parser, and not the property parser, and this is in `CSSPropertyParserConsumer+Ident.h`

@weinig Do you have an idea how to share this?

tuankiet65 replied:
Yeah I did see `consumeCustonIdentRaw` in `CSSPropertyParserConsumer+Ident.h`, not sure how they can share though

tuankiet65 replied:
I just blindly include `CSSPropertyParserConsumer+Ident.h`, not sure if it's allowed (in the sense of clean code)

nt1m commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -277,6 +277,46 @@ static FixedVector<PossiblyQuotedIdentifier> consumeLangArgumentList(CSSParserTo
+    return token.value().toAtomString();
+}
+
+static std::optional<FixedVector<PossiblyQuotedIdentifier>> consumeCommaSeparatedCustomIdentsList(CSSParserTokenRange& range)
```
```suggestion
static std::optional<FixedVector<PossiblyQuotedIdentifier>> consumeCommaSeparatedCustomIdentList(CSSParserTokenRange& range)
```

nt1m commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -277,6 +277,46 @@ static FixedVector<PossiblyQuotedIdentifier> consumeLangArgumentList(CSSParserTo
+            return std::nullopt;
+
+        customIdents.append({ *maybeItem, false });
+    }
```
write this as a `do { } while (...)` to avoid duplicating logic. Something like this:

```
     do {
         auto ident = consumeCustomIdent(range);
         if (!ident)
             return std::nullopt;
         customIdents.append({ *maybeItem, false });
     } while (consumeCommaIncludingWhitespace(args));
```

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -376,17 +376,17 @@ static void appendPseudoClassFunctionTail(StringBuilder& builder, const CSSSelec
 }
 
-static void appendLangArgument(StringBuilder& builder, const PossiblyQuotedIdentifier& langArgument)
+static void appendPossiblyQuotedIdentifier(StringBuilder& builder, const PossiblyQuotedIdentifier& argument)
```
nit: I would name it `identifier`. `argument` is quite generic.

nt1m commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -853,6 +876,13 @@ std::unique_ptr<MutableCSSSelector> CSSSelectorParser::consumePseudo(CSSParserTo
+            auto maybeTypesList = consumeCommaSeparatedCustomIdentList(block);
+            if (!maybeTypesList)
+                return nullptr;
+            selector->setArgumentList(WTFMove(*maybeTypesList));
```
nit: no need for the `maybe` prefix, the null check is self-explanatory imo

nt1m commented on Source/WebCore/cssjit/SelectorCompiler.cpp:
```diff
@@ -1225,6 +1225,7 @@ static inline FunctionType addPseudoClassType(const CSSSelector& selector, Selec
     case CSSSelector::PseudoClass::WebKitDrag:
     case CSSSelector::PseudoClass::Has:
     case CSSSelector::PseudoClass::State:
+    case CSSSelector::PseudoClass::ActiveViewTransitionType:
```
Can you file a bug about adding CSSJIT support?

tuankiet65 replied:
https://bugs.webkit.org/show_bug.cgi?id=278189
I'll add it to the code too

tuankiet65 committed:
[view-transitions] Support parsing and checking :active-view-transition-type()
https://bugs.webkit.org/show_bug.cgi?id=277779
rdar://133610087

Reviewed by Tim Nguyen.

This patch adds support for parsing and checking the pseudo class
:active-view-transition-type(). This patch also modifies ViewTransitionTypeSet
to invalidate the stylesheet if a new type is added/removed. With this, support
for view transition types is completed, and is now enabled by default.

The WPT test changes are already upstreamed here:
https://github.com/web-platform-tests/wpt/commit/85b16fcfb299c9d6ff85dfe16beacda7fcb609d2

* LayoutTests/TestExpectations: Don't skip view transition type tests, now that they're passing.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-matches-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-matches-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-matches.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved-mutation-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved-mutation.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-reserved.html:
* LayoutTests/platform/glib/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Enable ViewTransitionTypesEnabled by default
* Source/WebCore/css/CSSPseudoSelectors.json: added pseudo class :active-view-transition-type.
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::appendPossiblyQuotedIdentifier): renamed from appendLangArgument.
(WebCore::appendCommaSeparatedPossiblyQuotedIdentifierList): renamed from appendLangArgumentList.
(WebCore::CSSSelector::selectorText const): added code to serialize :active-view-transition-type.
(WebCore::appendLangArgument): Deleted.
(WebCore::appendLangArgumentList): Deleted.
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const): added code to check :active-view-transition-type.
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesActiveViewTransitionTypePseudoClass): added function to check :active-view-transition-type.
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h: added flag for parsing :active-view-transition-type.
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::consumeCommaSeparatedCustomIdentList): added function to parse a list of comma-separated custom idents (used by :active-view-transition-type)
(WebCore::CSSSelectorParser::consumePseudo): added code to parse :active-view-transition-type.
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h: added flag for parsing :active-view-transition-type.
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType): mark :active-view-transition-type as not JIT compilable.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setActiveViewTransition): when invalidating style, also invalidate :active-view-transition-type.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/ViewTransition.h:
(WebCore::ViewTransition::setTypes): use WTFMove.
* Source/WebCore/dom/ViewTransitionTypeSet.cpp:
(WebCore::ViewTransitionTypeSet::clearFromSetLike): invalidate style on clearing.
(WebCore::ViewTransitionTypeSet::addToSetLike): invalidate style on adding new types.
(WebCore::ViewTransitionTypeSet::removeFromSetLike): invalidate style on deleting types.
(WebCore::ViewTransitionTypeSet::hasType const): renamed from has()
(WebCore::ViewTransitionTypeSet::has const): Deleted.
* Source/WebCore/dom/ViewTransitionTypeSet.h:

Canonical link: https://commits.webkit.org/282344@main
